Content.Shared/_Impstation/CosmicCult/SharedMonumentSystem.cs

```diff
@@ -0,0 +1,163 @@
+public sealed class SharedMonumentSystem : EntitySystem
+{
+    [Dependency] private readonly SharedUserInterfaceSystem _uiSystem = default!;
+    [Dependency] private readonly SharedAppearanceSystem _appearance = default!;
+    [Dependency] private readonly SharedMapSystem _map = default!;
+    [Dependency] private readonly IGameTiming _timing = default!;
+    [Dependency] private readonly IPrototypeManager _prototype = default!;
+    [Dependency] private readonly SharedActionsSystem _actions = default!;
+    public override void Initialize()
+    {
+        base.Initialize();
+        SubscribeLocalEvent<MonumentComponent, BoundUIOpenedEvent>(OnUIOpened);
+        SubscribeLocalEvent<MonumentComponent, GlyphSelectedMessage>(OnGlyphSelected);
+        SubscribeLocalEvent<MonumentComponent, GlyphRemovedMessage>(OnGlyphRemove);
+        SubscribeLocalEvent<MonumentComponent, InfluenceSelectedMessage>(OnInfluenceSelected);
+        SubscribeLocalEvent<MonumentComponent, PreventCollideEvent>(OnPreventCollide);
+    }
+    public override void Update(float frameTime)
+    {
+        base.Update(frameTime);
+        var query = EntityQueryEnumerator<MonumentTransformingComponent>();
+        while (query.MoveNext(out var uid, out var comp))
+        {
+            if (_timing.CurTime < comp.EndTime)
+                continue;
+            _appearance.SetData(uid, MonumentVisuals.Transforming, false);
+            RemComp<MonumentTransformingComponent>(uid);
+        }
+    }
+    private void OnPreventCollide(EntityUid uid, MonumentComponent comp, ref PreventCollideEvent args) // Ensures that Cultists can't walk through The Monument and allows non-cultists to walk through the space.
+    {
+        if (!HasComp<CosmicCultComponent>(args.OtherEntity) && !comp.HasCollision)
+            args.Cancelled = true;
+    }
+    private void OnUIOpened(Entity<MonumentComponent> ent, ref BoundUIOpenedEvent args)
+    {
+        if (!_uiSystem.IsUiOpen(ent.Owner, MonumentKey.Key) || !TryComp<ActivatableUIComponent>(ent, out var uiComp))
+            return;
+        if (ent.Comp.Enabled && TryComp<CosmicCultComponent>(uiComp.CurrentSingleUser, out var cultComp))
+        {
+            ent.Comp.UnlockedInfluences = cultComp.UnlockedInfluences;
+            ent.Comp.AvailableEntropy = cultComp.EntropyBudget;
+        }
+        else _uiSystem.CloseUi(ent.Owner, MonumentKey.Key); // based on the prior IF, this effectively cancels the UI if the user is either not a cultist, or the Finale is ready to trigger.
+        _uiSystem.SetUiState(ent.Owner, MonumentKey.Key, GenerateBuiState(ent.Comp));
+    }
+
+    #region UI listeners
+    private void OnGlyphSelected(Entity<MonumentComponent> ent, ref GlyphSelectedMessage args)
+    {
+        // TODO: this needs checks for tier, or mote cost, or whatever you want to do here
+
+        ent.Comp.SelectedGlyph = args.GlyphProtoId; // not sure SelectedGlyph is needed for anything? keeping it here in case
```

Comment:
It's used to make the selected glyph's button be pressed in MonumentMenu.xaml.cs